File: vscode-extension/src/rules/config/placeholderEnvValuesRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaceholderEnvValuesRule = void 0;
const finding_1 = require("../../models/finding");
const SENSITIVE_KEYS = new Set(['SUPABASE_URL', 'SUPABASE_ANON_KEY']);
class PlaceholderEnvValuesRule {
    constructor() {
        this.code = 'placeholder-env-value';
    }
    evaluate(context) {
        if (!context.usesSupabase) {
            return [];
        }
        const findings = [];
        for (const entry of context.envEntries) {
            if (!SENSITIVE_KEYS.has(entry.key)) {
                continue;
            }
            if (entry.file.isEnvTemplateFile) {
                continue;
            }
            if (!this._looksLikePlaceholder(entry.value)) {
                continue;
            }
            findings.push(new finding_1.Finding({
                severity: finding_1.FindingSeverity.low,
                confidence: finding_1.FindingConfidence.high,
                category: finding_1.FindingCategory.config,
                code: this.code,
                message: `${entry.key} appears to be a placeholder value in ${entry.file.name}`,
                fix: 'Replace the placeholder with the real value from your Supabase project settings before running the app.',
                risk: 'Using placeholder configuration values will cause network requests or authentication to fail.',
                filePath: entry.file.relativePath,
                line: entry.line,
            }));
        }
        return findings;
    }
    _looksLikePlaceholder(value) {
        if (!value) {
            return true;
        }
        const normalized = value.toLowerCase().trim();
        if (['changeme', 'change_me', 'replace_me', 'replace-me', 'todo',
            'your_key_here', 'your-key-here', 'your_url_here', 'your-url-here'].includes(normalized)) {
            return true;
        }
        if (/^<[^>]+>$/.test(value.trim())) {
            return true;
        }
        if (/^[A-Z][A-Z0-9_]+$/.test(value.trim())) {
            return true;
        }
        return normalized.includes('your-') || normalized.includes('your_') ||
            normalized.includes('placeholder') || normalized.includes('example') ||
            normalized.includes('xxx');
    }
}
exports.PlaceholderEnvValuesRule = PlaceholderEnvValuesRule;

File: lib/src/rules/config/placeholder_env_values_rule.dart
import '../../models/finding.dart';
import '../../models/project_context.dart';
import '../../rule.dart';

/// Detects `.env` entries whose values are clearly placeholder text rather
/// than real secrets — e.g. `SUPABASE_ANON_KEY=YOUR_KEY_HERE`.
///
/// This catches the common mistake of committing an `.env` with scaffolded
/// values that were never replaced, which the `missing-env-vars` rule would
/// miss (the key is *present*, just wrong).
class PlaceholderEnvValuesRule extends Rule {
  const PlaceholderEnvValuesRule();

  @override
  String get code => 'placeholder-env-value';

  // Keys worth checking — only those the scanner actually cares about.
  static const _sensitiveKeys = {'SUPABASE_URL', 'SUPABASE_ANON_KEY'};

  @override
  List<Finding> evaluate(ProjectContext context) {
    if (!context.usesSupabase) {
      return const [];
    }

    final findings = <Finding>[];

    for (final entry in context.envEntries) {
      if (!_sensitiveKeys.contains(entry.key)) {
        continue;
      }
      if (entry.file.isEnvTemplateFile) {
        // Template files (.env.example etc.) are expected to have placeholders.
        continue;
      }
      if (!_looksLikePlaceholder(entry.value)) {
        continue;
      }

      findings.add(
        Finding(
          severity: FindingSeverity.low,
          confidence: FindingConfidence.high,
          category: FindingCategory.config,
          code: code,
          message:
              '${entry.key} appears to be a placeholder value in ${entry.file.name}',
          fix:
              'Replace the placeholder with the real value from your Supabase project settings before running the app.',
          risk:
              'Using placeholder configuration values will cause network requests or authentication to fail.',
          filePath: entry.file.relativePath,
          line: entry.line,
        ),
      );
    }

    return findings;
  }

  bool _looksLikePlaceholder(String value) {
    if (value.isEmpty) {
      return true;
    }
    final normalized = value.toLowerCase().trim();
    // Explicit placeholder keywords.
    if (normalized == 'changeme' ||
        normalized == 'change_me' ||
        normalized == 'replace_me' ||
        normalized == 'replace-me' ||
        normalized == 'todo' ||
        normalized == 'your_key_here' ||
        normalized == 'your-key-here' ||
        normalized == 'your_url_here' ||
        normalized == 'your-url-here') {
      return true;
    }
    // Angle-bracket templates: <SUPABASE_ANON_KEY>, <your-project-ref>.
    if (RegExp(r'^<[^>]+>$').hasMatch(value.trim())) {
      return true;
    }
    // ALL_CAPS env-var-style like YOUR_KEY_HERE or SUPABASE_URL (the key name
    // itself being used as the value is a clear copy-paste artifact).
    if (RegExp(r'^[A-Z][A-Z0-9_]+$').hasMatch(value.trim())) {
      return true;
    }
    // Contains 'your-', 'placeholder', 'example', 'xxx'.
    return normalized.contains('your-') ||
        normalized.contains('your_') ||
        normalized.contains('placeholder') ||
        normalized.contains('example') ||
        normalized.contains('xxx');
  }
}
